Adds python module and OpenCL support

Bidski opened the pull request.

Commit:
Adds python module support

Commit:
Adds ability for python modules to emit vectors of messages

Updates pybind syntax to suppress warnings about int

Commit:
Ensures pybind exists for messaging library

Removes copying of nuclear.py (this is handled in NUClearModule.cmake now)

Commit:
Adds missing sys.argv

Commit:
Upgrades module generation to include the generation of a python template

Commit:
Adds python and dependencies to toolchain

Commit:
Adds pybind11 to toolchain

Commit:
Sets PYTHONPATH in generated cmake files

Updates syntax of generated cmake files

Commit:
Adds new environment variables to installer script

Improves conditional build handling

Commit:
Implements zpreztorc regex as a sed command

This should work better than the original method

Commit:
Updates toolchain version

Commit:
Adds pyyapf for python formatting

Commit:
Adds functionality to intall python modules

Commit:
Formats all python files

Commit:
Formats all python files

Commit:
Adds ability for cmake to check the installation of python modules

Commit:
Adds toolchain environment to travis build

Commit:
Moves toolchain environment into zshrc

Commit:
Builds an unoptimised python message library

This is needed so that older CPUs can build python messages

Commit:
Installs pyyapf in travis

Commit:
Adds python3-pip to travis

Commit:
Fixes python formatting

Commit:
Prevents pip from trying to install packages multiple times

Commit:
Allow libraries to find the python include files

Commit:
Updates boost version

Commit:
Merge branch 'master' into biddulph/python-modules

Commit:
Fixes include flags

Commit:
Adds OpenCL SDK

Commit:
Adds OpenCL support

Commit:
Adds an OpenCL module

Commit:
Adds a python module

Commit:
Fixes role generation

Python header files are generated in the build folder, we need to search there rather than the source folder for them

Commit:
Generates stubs classes for python messages

The c++ message shared library can not be executed on all systems due to compile-time optimisations. The stubs classes provides the python build system with enough information to allow the build to finish

Commit:
Updates toolchain version

Commit:
Fixes travis build file to use stages correctly

Commit:
Revert "Fixes formatting"

This reverts commit 96b184ce17c8185fd97469d5e13f9c820162e3aa.

Commit:
Fixes formatting

Commit:
Corrects include_paths for generated python messages

Commit:
Fixes python message dependencies

Commit:
Changes formatting script to search all directories

This now matches where the travis fomatting validation script searches

Commit:
Revert "Changes formatting script to search all directories"

This reverts commit 0d165540282c800614bd7911ceeb42d30e1c4b7b.

Commit:
Fixes formatting

Commit:
Revert "Fixes formatting"

This reverts commit 3498744996e65af6c77f1ae7091c9092ddef714c.

Commit:
Revert "Fixes formatting"

This reverts commit 7db2c9d84c1bb6b613ceebc0f521ed331d1ad7d9.

Commit:
Fixes formatting

Commit:
Merge branch 'master' into biddulph/python-modules

Commit:
Removes formatting

Commit:
Revert "Formats all python files"

This reverts commit 87b6430e35644455076bc1ba3754eda8401f7886.

Commit:
Revert "Formats all python files"

Commit:
Revert of "Formats all python files"

Commit:
Reverts last of the formatting changes

TrentHouliston commented on nuclear/module/NUClearModule.cmake:
```diff
@@ -87,24 +98,35 @@ FUNCTION(NUCLEAR_MODULE)
 
+        SET(ENV{PYTHONPATH} "${PYTHONPATH}:${PYTHON_SITE_PACKAGES}")
+        SET(ENV{NUCLEAR_MODULE_DIR} "${PROJECT_SOURCE_DIR}/${NUCLEAR_MODULE_DIR}")
+        SET(ENV{LD_LIBRARY_PATH} "/nubots/toolchain/${PLATFORM}/lib:/nubots/toolchain/lib")
```
This has to be smarter than this, This code is shared between all NUClear projects

Bidski replied:
Hmmm .... this will take some thought

TrentHouliston commented on nuclear/module/NUClearModule.cmake:
```diff
@@ -87,24 +98,35 @@ FUNCTION(NUCLEAR_MODULE)
-                PYTHONPATH="${PROJECT_BINARY_DIR}/python/nuclear/"
+                PYTHONPATH="${PROJECT_BINARY_DIR}/python/nuclear:${PYTHONPATH}:${PYTHON_SITE_PACKAGES}"
                 NUCLEAR_MODULE_DIR="${PROJECT_SOURCE_DIR}/${NUCLEAR_MODULE_DIR}"
+                LD_LIBRARY_PATH="/nubots/toolchain/${PLATFORM}/lib:/nubots/toolchain/lib"
```
Once again nubots/toolchain won't work here

TrentHouliston commented on nuclear/tools/module.py:
```diff
@@ -107,15 +133,20 @@ class {className} : public NUClear::Reactor {{
+
+    return template.format(
+        define='{}_H'.format('_'.join([p.upper() for p in parts])),
+        className=parts[-1],
```
A wild camelCase appeared

Bidski replied:
Pretty sure these are your fault, I just formatted the code lol

TrentHouliston commented on nuclear/tools/module.py:
```diff
@@ -132,14 +163,19 @@ def generate_cpp(parts):
+    )
+
+    return template.format(
+        className=parts[-1],
```
More camelCase

TrentHouliston commented on nuclear/module/NUClearModule.cmake:
```diff
@@ -171,9 +193,14 @@ FUNCTION(NUCLEAR_MODULE)
+        LIST(APPEND NUCLEAR_MODULE_DATA_FILES ${output_file})
     ENDFOREACH(data_file)
 
+    # Remove duplicate entries from NUCLEAR_MODULE_DATA_FILES
```
This comment is wrong

TrentHouliston commented on nuclear/module/python/nuclear.py:
```diff
@@ -152,11 +158,29 @@ def include_paths(self):
+def bind_emit(self, msg):
+    # Iterable type
+    if is_sequence(msg):
+        return msg[0]._emit_vector(self._reactor_ptr, msg)
```
Still don't know how I feel about this emit vector thing, We're trying to get rid of it in the rest of the code too

Bidski replied:
Well until it is completely removed from the rest of code this needs to stay

TrentHouliston commented on nuclear/tools/module.py:
```diff
@@ -154,22 +190,45 @@ def generate_readme(parts):
+    )
+
+    return template.format(
+        className=parts[-1], classNameTitle=len(parts[-1]) * '=', closeNamespace='\n'.join(['}' for x in parts[:-1]])
```
Even more!!

TrentHouliston commented on nuclear/tools/module.py:
```diff
@@ -154,22 +190,45 @@ def generate_readme(parts):
+                # Constructor for {className}
+
+            @on(Configuration('{className}.yaml'))
+            def {className}_configfuration(self, config):
```
Did you actually get this code to work?

Bidski replied:
No, I should probably comment this out here for now. We need to create more bindings in general though (extensions and utilities)

TrentHouliston reviewed (changes requested):
Some changes

Commit:
Removes camel casing

Commit:
Disables python configuration

Extension bindings need to be implemented

Commit:
Fixes comment

Commit:
Random updates

Commit:
Random updates

Commit:
Changes protobuf to use cmake build system

Commit:
Removes commented code

Commit:
Removes redundant code

Commit:
Fixes python message generation code

Commit:
Removes FSwatch version check

Commit:
Updates pybind contructor calls

Commit:
Makes opencl install self-recoverable

Commit:
Organises apt ppas and packages

Commit:
Installs unzip as part of build tools

Commit:
Adds travis-ci upgrade ppa for newer gettext version

Commit:
Prevents protobuf from rebuilding unnecessarily

Commit:
Merge branch 'master' into biddulph/python-modules

Commit:
Formatting ....

Commit:
Merge remote-tracking branch 'origin/biddulph/python-modules' into biddulph/python-modules

Commit:
Installs newer cmake in travis

Commit:
Installs newer cmake in travis

Commit:
Merge remote-tracking branch 'origin/biddulph/python-modules' into biddulph/python-modules

Commit:
Travis debugging

Commit:
Travis debugging

Commit:
Merge remote-tracking branch 'origin/biddulph/python-modules' into biddulph/python-modules

Commit:
Merge remote-tracking branch 'origin/biddulph/python-modules' into biddulph/python-modules

Commit:
Merge remote-tracking branch 'origin/biddulph/python-modules' into biddulph/python-modules

Commit:
Modifies script extension to use a protobuf message

Primarily to allow python modules to execute scripts, however python needs to subsumption for this to actually work

Commit:
Adds a hack module to get python around subsumption

Commit:
Fixes module generation

Commit:
Adds missing message file

Commit:
Fixes bad merge

Commit:
Fixes travis and puppet bs

Commit:
Fixes travis and puppet bs

Commit:
Merge remote-tracking branch 'origin/biddulph/python-modules' into biddulph/python-modules

Commit:
Fixes generated python module file

Commit:
Fixes module language specification

Commit:
Adds module formatting for cpp files

Relies on clang-format-6, but should fail cleanly if its not installed

Commit:
Merge remote-tracking branch 'origin/biddulph/python-modules' into biddulph/python-modules

Commit:
Merge branch 'master' into biddulph/python-modules

KipHamiltons commented:
@Bidski does this PR have stuff worth porting to the current codebase?

Bidski commented:
The python module support needs a lot more work to be properly usable (utilities need python wrappings). The Opencl support is mostly just an extension so you can do `on<OpenCL>` which is only useful if we think we will be implementing more OpenCL-based modules

Bidski commented:
We also need to see if we can figure out how to wrap extensions in python as well

KipHamiltons commented:
> The python module support needs a lot more work to be properly usable (utilities need python wrappings). 

what would that enable us to do?

Bidski commented:
> what would that enable us to do?

You mean apart from being able to write modules in python that integrate nicely with the C++ modules?

KipHamiltons commented:
No, that answers my question directly. That would be pretty nice!!!
